src/conf_mode/container.py

```diff
@@ -0,0 +1,176 @@
+                image = container_config['image']
+                repo = cmd(f'docker image ls --format ' + '{{.Repository}}' + f' --filter \"reference={image}\"')
+                # Check if image was pulled and present in local images
+                if image != repo:
+                    raise ConfigError(f'Pull image for container "{name}" is required!\nUse: run add container image {image}')
+
+            # If 'allow-host-networks' or 'network' not set.
+            if 'allow_host_networks' not in container_config and 'network' not in container_config:
+                raise ConfigError(f'"Network" or "allow-host-networks" for container "{cont}" is required!')
+
+            # If set both parameters for networks (host and user-defined). We require only one.
+            if 'allow_host_networks' in container_config and 'network' in container_config:
+                raise ConfigError(f'"allow-host-networks" and "network" for "{cont}" cannot be both configured at the same time!')
+
+    # Add new network
+    if 'network' in container:
+        for net in container['network']:
+            # If ipv4-prefix not defined for user-defined network
+            if 'ipv4_prefix' not in container['network'][net]:
+                raise ConfigError(f'IPv4 prefix for network "{net}" is required!')
+
+    return None
+
+def generate(container):
+    # bail out early - looks like removal from running config
+    if not container:
+        return None
+
+    return None
+
+def apply(container):
+    # Delete old networks if needed
+    if 'net_remove' in container:
+        for net in container['net_remove']:
+            call(f'docker network rm {net}')
+
+    # Delete old containers if needed. We can't delete running container
+    if 'container_remove' in container:
+        for name in container['container_remove']:
+            call(f'docker stop {name}')
+            call(f'docker rm {name}')
+
+    if not 'configured' in container:
+        # Stop docker.service, we should stop docker.socket, otherwise the service won't stop
+        call('systemctl stop docker.socket')
+        return
+
+    # Add network
+    if 'network' in container:
+        for net in container['network']:
+            # Check if the network has already been created
+            if len(cmd(f'docker network ls  --filter "name=^{net}$" -q')) == 0:
+                prefix = container['network'][net]['ipv4_prefix']
+                if container['network'][net]['ipv4_prefix']:
+                    # Create user-defined bridge network
+                    try:
+                        call(f'docker network create {net} --subnet={prefix}')
+                    except:
+                        print(f'Can\'t add network {net}')
+
+    # Add container
+    if 'name' in container:
+        for name in container['name']:
+            # Check if the container has already been created
+            if len(cmd(f'docker ps -a --filter "name=^{name}$" -q')) == 0:
```

Comment:
I believe we should find a way to check if a container is running that doesn't rely on outputs that aren't supposed to be machine-readable. That's my only real objection.

Reply:
https://docs.docker.com/engine/api/sdk/

I hope this information may help implementers. This is a very good feature. I expect it to complete

Reply:
According to this Python SDK, in fact, we do not need to execute any operating system commands to run docker

Reply:
This is beneficial to python without any harm. If possible, consider using python's `docker` library

Why pass the command line to the operating system when the python library can be used directly?

Reply:
@jack9603301 It has not yet been decided whether it will be docker or not. The scheme is simplified as much as possible (number of pkgs/commands/etc), In order to understand whether it makes sense to further get the dependencies of additional API.
Of course, using native/additional/python API will certainly be better and more correct.

Reply:
No matter what scheme is used, I expect VyOS to support containers